ublox/src/ubx_packets/packets.rs

```diff
@@ -1939,111 +1941,123 @@ struct RxmRtcm {
 #[ubx(class = 0x10, id = 0x02, max_payload_len = 1240)]
 struct EsfMeas {
     time_tag: u32,
+    flags: u16,
+    id: u16,
+    #[ubx(
+        map_type = EsfMeasDataIter,
+        from = EsfMeasDataIter::new,
+        size_fn = data_len,
+        is_valid = EsfMeasDataIter::is_valid,
+        may_fail,
+    )]
+    data: [u8; 0],
     #[ubx(
-        map_type = EsfMeasInfo,
-        from = EsfMeasInfo::new,
-        flatten,
-        flat_fields = ["flags", "u16", "id", "u16", "data", "U32Iter"]
+        map_type = Option<u32>,
+        from = EsfMeas::calib_tag,
+        size_fn = calib_tag_len,
     )]
-    //, "calib_tag", "Option<u32>"
-    info: [u8; 0],
+    calib_tag: [u8; 0],
 }
 
-#[derive(serde::Serialize, Debug)]
-pub struct EsfMeasInfo<'a> {
-    pub flags: u16,
-    pub id: u16,
-    pub data: U32Iter<'a>,
-    pub calib_tag: Option<u32>,
+impl EsfMeas {
+    fn calib_tag(bytes: &[u8]) -> Option<u32> {
+        bytes.try_into().ok().map(u32::from_le_bytes)
+    }
 }
 
-#[ubx_iter]
-pub struct U32Iter<'a>(core::slice::ChunksExact<'a, u8>);
+impl<'a> EsfMeasRef<'a> {
+    fn data_len(&self) -> usize {
+        ((self.flags() >> 11 & 0x1f) as usize) * 4
```

Comment:
ohhhhhh i see what u did, like assuming the function flags is generated inside packets.rs and reserving the size except it doesn't get consumed (so variable doesn't have to be declared at the end like before) and basically just slicing the bytes in the generated code?

thought this way was going to be more complicated than this

Reply:
also, just to confirm - are these functions always generated at compile time only compiled once like cargo build and release would not run expanding macros (no compromise with speed etc)? 

is this just like lombok in java except its not as "hacky" in rust and more of standard?

Reply:
> ohhhhhh i see what u did, like assuming the function flags is generated inside packets.rs and reserving the size except it doesn't get consumed (so variable doesn't have to be declared at the end like before) and basically just slicing the bytes in the generated code?

yup! I think the key was figuring we could add functions to the Ref structs outside of the code gen. Then yeah you just have to check that there are enough bytes in the slice to use whatever methods you added 

> also, just to confirm - are these functions always generated at compile time only compiled once like cargo build and release would not run expanding macros (no compromise with speed etc)?
>
> is this just like lombok in java except its not as "hacky" in rust and more of standard?

Not familiar w lombok but yeah looks similar. Proc macros (https://doc.rust-lang.org/reference/procedural-macros.html) are actually rustc compiler extensions. They get invoked/expanded before the real code gen happens. So if you run cargo build or check back to back twice it should only generate once. If you change the proc macro library (or any code you are calling it on) then it should rebuild

Reply:
If you run build then build --release it would also build twice I think, afaik release and debug builds are entirely separate